### DISCUSSION TITLE
Is it possible to generate metadata dynamically before render?

### DISCUSSION THREAD

**Discussion by benabel** [quarto-dev/quarto-cli#9895] [Q&A] [answered]
### Description

On my previous websites using Gatsby or eleventy I was able to generate metadata at build time.

For example: extracting information from file path: `.../1-Introduction to statistics/exercise/index.md` let me get this metadata:

```
title: Introduction to statistics
order: 1
type: exercise
permalink: /introduction-to-statistics/exercise
```

This is handled by an awesome feature called _data cascade_ in eleventy see the [doc](https://www.11ty.dev/docs/data-computed/#using-javascript) here.

Is it possible to do something similar in quarto?

**Comment by mcanouil** [comment]
Yes, you can use pre-render script and the 1.5 pre-release.

- https://quarto.org/docs/projects/scripts.html#pre-and-post-render
- https://quarto.org/docs/prerelease/1.5/pre-render.html

Note that `order`, `type`, `permalink` from your example are not Quarto YAML options.
See the documentation.

**Comment by benabel** [comment]
Thank you @mcanouil for these directions , I'll investigate on pre-render scripts with 1.5 version. Is there any way to change the url to a page dynamically or is it always computed from the file path?

**Reply by mcanouil** [reply]
Look for aliases/redirection that what is closer to "permalink".

https://quarto.org/docs/websites/website-navigation.html#redirects

**Comment by benabel** [comment]
It seems that manipulating `permalink` is not possible, also I tried setting dynamically the title with a lua filter like this:

```lua
function Meta(meta)
    local title = meta.title
    meta["title"][1]["text"] = "New title"  
    return meta
end
```
Title is set correctly for the document **but breadcumb and navbars are not updated**. 

I imagine that the only way to handle this specific use is to modify the `qmd` files and save the changes, I think it would be nice to manipulate the metadata of files dynamically earlier. I close the discussion. Thanks for your time and this incredible program.

**Reply by mcanouil** [reply, marked as answer]
Some metadata are needed to actually build the structure and for Quarto to know what to build and where.
Your filter to modify the title is a bit odd but if it works.
Use `format: native` and look at the AST representation of the document when developing filter, it helps.

As I said, `permalink` does not exist, but you can define redirection in your YAML frontmatter.
I don't know if it will work to generate them dynamically in a pre-render script though.
